src/bromwhich_contour.jl

```diff
@@ -0,0 +1,138 @@
+"""
+    hyperbola(f::Function, t::AbstractFloat; N::Int = 16)
+
+Evaluate the inverse Laplace transform of `f` at the point `t` by approximating the Bromwhich integral with a hyperbola contour.
+A real valued time-domain signal is assumed with the integral being calculated by a computable series applying the midpoint rule.
+
+The contour is defined by several parameters. μ controls the extreme nodes with a larger parameter moving the outlier nodes further into left half plane.
+h is the uniform node spacing with N being the number of nodes on the contour. The parameters depend on both t and N such that the function must be computed N times for each t.
+This is very inefficient especially if f is computational expensive, so this approach is recommended when only a few values of t are needed.
+However, this method will provide more accurate results at the cost of computational time compared to fixed contour approaches.
+
+The number of nodes, N, defaults to 16. The optimal N is case dependent and arbitrarily increasing N will not result in more accurate results. Smaller values of N will be faster.
+
+Bromwhich contour approaches should only be applied for t > 0. We are aiming to optimize the convergence as N -> inf so our parameters are ~ N / t.
+In other words we want the Bromwhich integral to converge rapidly, and this can be done by starting and ending our integration path in the left hand plane causing exp(st) to decay.
+If t = 0, the exponential factor does not cause good convergence. If you want to evaluate for t = 0, you should try a small positive value like 1e-30.
+
+# Example
+
+```jldoctest
+julia> InverseLaplace.hyperbola(s -> 1/(s + 1), 2.0)
+0.13533528323665164
+"""
+function hyperbola(f::Function, t::AbstractFloat; N = 16)
+    a =  zero(Complex{eltype(t)})
+    N = convert(eltype(t), N)
+    h = 1.081792140 / N
+    for k in 0:N-1
+        sk = s((k + 1/2) * h, N, t)
+        dsk = ds((k + 1/2) * h, N, t)
+        a += f(sk) * exp(sk * t) * dsk
+    end
+    return imag(a) * h / pi
+end
+
+# loop through an entire time array with multithreads
+function hyperbola(f::Function, t::AbstractArray; N = 16)
+    out = similar(t)
+    Threads.@threads for ind in eachindex(t)
+        out[ind] = hyperbola(f, t[ind], N = N)
+    end
+    return out
+end
+
+#### fixed integration path (can precompute) ####
+
+# get the fixed integration components
+function hyper_coef(N, t::AbstractArray; ϕ = 1.09)
+    A = acosh(((π - 2 * ϕ) * t[end] / t[1] + 4 * ϕ - π) / ((4 * ϕ - π) * sin(ϕ)))
+    μ = (4 * π * ϕ - π^2) * N / t[end] / A
+    h = A / N
+    return μ, h
+end
+
+# compute the fixed hyperbola contour
+s_fixed(θ, μ; ϕ = 1.09) = μ + im * μ * sinh(θ + im * ϕ)
+ds_fixed(θ, μ; ϕ = 1.09) = im * μ * cosh(θ + im * ϕ)
+
+# compute the function values over the fixed contour nodes
+function fixed_sk(f::Function, N, t::AbstractArray)
+    μ, h = hyper_coef(N, t)
+    a = zeros(Complex{eltype(h)}, Int(N))
```

Comment:
Hmmm I always go back and forth on the best way to handle this. But we are allocating quite a bit within this function that the user can't do anything about it. And then we call `Threads` which greatly increases the allocations as well. If the user is calling this within a loop they'll see a lot of allocations. Wondering if she provide some pre cache option here....

Reply:
You should probably used `typeof` rather than `eltype` for things that aren't containers. Although `eltype` works. Maybe because floats and ints are iterable.

It's common to provide `fixed_sk!` that takes a buffer as input and then `fixed_sk` just does this allocation and calls `fixed_sk!`.

I don't know the best way to handle threads; you can provide a keyword argument. You can also have a conditional that turns on threads only if the work is big enough.